src/renderer/src/contexts/ModalRouter.tsx

```diff
@@ -0,0 +1,137 @@
+import { THistory, TModalRouterContextValue, TModalRouterProviderProps } from '@renderer/@types/modal'
+import { twMerge } from 'tailwind-merge'
+
+type ModalProps = {
+  children: React.ReactNode
+  onClose: (name: string) => void
+  history: THistory
+}
+
+const Modal = ({ children, onClose, history }: ModalProps) => {
+  const { closeOnEsc, closeOnOutsideClick, className, ...props } = history.props ?? {}
+
+  const containerRef = useRef<HTMLDivElement>(null)
+
+  const handleContentClick = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
+    event.stopPropagation()
+  }
+
+  const handleModalClick = () => {
+    if (closeOnOutsideClick === false) {
+      return
+    }
+
+    onClose(history.name)
+  }
+
+  const handleModalKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
+    if (closeOnEsc === false || event.key !== 'Escape') {
+      return
+    }
+
+    onClose(history.name)
+  }
+
+  useEffect(() => {
+    containerRef.current?.focus()
+  }, [])
+
+  return (
+    <div
+      className={twMerge(
+        'fixed left-0 top-0 w-screen h-screen flex justify-end bg-gray-black/50 backdrop-blur-sm',
+        className
+      )}
+      onClick={handleModalClick}
+      onKeyDown={handleModalKeyDown}
+      ref={containerRef}
+      tabIndex={0}
+      {...props}
+    >
+      <div onClick={handleContentClick} onKeyDown={handleModalKeyDown} className="h-full">
+        {children}
+      </div>
+    </div>
+  )
+}
+
+const ModalPortal = (props: ModalProps) => {
+  const modalRoot = document.querySelector('body') as HTMLBodyElement
+  return createPortal(<Modal {...props} />, modalRoot)
+}
+
+export const ModalRouterContext = createContext<TModalRouterContextValue>({} as TModalRouterContextValue)
+
+export const ModalRouterProvider = ({ routes, children }: TModalRouterProviderProps) => {
+  const [history, setHistory] = useState<THistory[]>([])
+
+  const setNonRepeatedHistoryElement = useCallback((newHistory: THistory) => {
+    setHistory(prevState => {
+      const prevStateFiltered = prevState.filter(item => item.name !== newHistory.name)
+
+      return [...prevStateFiltered, newHistory]
+    })
+  }, [])
+
+  const navigate = useCallback(
+    (name: string | number, options?: Omit<THistory, 'name'>) => {
+      if (typeof name === 'string') {
+        const routeExist = routes.some(route => route.name === name)
+        if (!routeExist) {
+          throw new Error(`Route not found: ${name}`)
+        }
+
+        return setNonRepeatedHistoryElement({ name, ...options })
+      }
+
+      if (name >= 0) {
+        const routeElement = routes[name]
+        if (!routeElement) {
+          throw new Error(`Route not found: ${name}`)
+        }
+
+        return setNonRepeatedHistoryElement(routeElement)
+      }
+
+      setHistory(prevState => prevState.slice(0, name))
+      return
+    },
+    [setNonRepeatedHistoryElement, routes]
+  )
+
+  const getElementByHistory = (historyElement: THistory) => {
```

Comment:
I think this can be a callback, right?

Reply:
Since we are only using it in the rendering phase, there is no need to create a callback for it as no react hook depends on it. Furthermore, React doesn't let you call a callback inside a loop https://react.dev/reference/react/useCallback#i-need-to-call-usememo-for-each-list-item-in-a-loop-but-its-not-allowed.